cirq-aqt/cirq_aqt/aqt_device.py

```diff
@@ -218,13 +223,148 @@ def simulate_samples(self, repetitions: int) -> cirq.Result:
+            oneq_gates_duration: The maximum duration of a single qubit
+            operation.
+            qubits: Qubits on the device, identified by their x location.
+
+        Raises:
+            TypeError: If not all the qubits supplied are `cirq.LineQubit`s.
+        """
+        self._measurement_duration = value.Duration(measurement_duration)
+        self._twoq_gates_duration = value.Duration(twoq_gates_duration)
+        self._oneq_gates_duration = value.Duration(oneq_gates_duration)
+        if not all(isinstance(qubit, devices.LineQubit) for qubit in qubits):
+            raise TypeError(
+                "All qubits were not of type cirq.LineQubit, instead were "
+                f"{set(type(qubit) for qubit in qubits)}"
+            )
+        self.qubits = frozenset(qubits)
+        self.gateset = AQTTargetGateset()
+
+        graph = nx.Graph()
+        graph.add_edges_from([(a, b) for a in qubits for b in qubits if a != b], directed=False)
+        self._metadata = devices.DeviceMetadata(self.qubits, graph)
+
+    @property
+    def metadata(self) -> devices.DeviceMetadata:
+        return self._metadata
+
+    def decompose_circuit(self, circuit: circuits.Circuit) -> circuits.Circuit:
+        return transformers.optimize_for_target_gateset(circuit, gateset=self.gateset)
+
+    def duration_of(self, operation):
+        if isinstance(operation.gate, ops.XXPowGate):
+            return self._twoq_gates_duration
+        if isinstance(
+            operation.gate, (ops.XPowGate, ops.YPowGate, ops.ZPowGate, ops.PhasedXPowGate)
+        ):
+            return self._oneq_gates_duration
+        if isinstance(operation.gate, ops.MeasurementGate):
+            return self._measurement_duration
+        raise ValueError(f'Unsupported gate type: {operation!r}')
+
+    def validate_gate(self, gate: ops.Gate):
+        if gate not in self.gateset:
+            raise ValueError(f'Unsupported gate type: {gate!r}')
+
+    def validate_operation(self, operation):
+        if not isinstance(operation, ops.GateOperation):
+            raise ValueError(f'Unsupported operation: {operation!r}')
+
+        self.validate_gate(operation.gate)
+
+        for q in operation.qubits:
+            if not isinstance(q, devices.LineQubit):
+                raise ValueError(f'Unsupported qubit type: {q!r}')
+            if q not in self.qubits:
+                raise ValueError(f'Qubit not on device: {q!r}')
+
+    def validate_circuit(self, circuit: circuits.AbstractCircuit):
+        super().validate_circuit(circuit)
+        _verify_unique_measurement_keys(circuit.all_operations())
+
+    def at(self, position: int) -> Optional[devices.LineQubit]:
+        """Returns the qubit at the given position, if there is one, else None."""
+        q = devices.LineQubit(position)
+        return q if q in self.qubits else None
+
+    def neighbors_of(self, qubit: devices.LineQubit) -> Iterable[devices.LineQubit]:
```

Comment:
This should likely move to metadata as well.

Reply:
These can be moved as well.